### ISSUE TITLE
[Bug]: SAM Callback Fails in MultiData MultiTask Setting With All `None` Param Gradients

### ISSUE THREAD

**Issue by melo-gonzo** [IntelLabs/matsciml#188]
### Expected behavior

I would either expect some of the optimizer parameter gradients to not be `None`, or for the SAM callback to handle the case when all of the parameter gradients are `None`.

### Actual behavior

When all the parameter gradients are none, [this line](https://github.com/IntelLabs/matsciml/blob/7375112f5f07a6e3ddd4f52ac25c661ff898aa84/matsciml/lightning/callbacks.py#L794) fails to compute because the list is empty.

Condensed Stack Trace:
```text
  File "/Users/carmelog/Projects/GitRepos/open-catalyst/public-repo/matsciml-fork/matsciml/lightning/callbacks.py", line 779, in on_before_optimizer_step
    org_weights = self._first_step(optimizer)
  File "/Users/carmelog/Projects/GitRepos/open-catalyst/public-repo/matsciml-fork/matsciml/lightning/callbacks.py", line 807, in _first_step
    scale = self.rho / (self._grad_norm(optimizer) + 1e-5)
  File "/Users/carmelog/Projects/GitRepos/open-catalyst/public-repo/matsciml-fork/matsciml/lightning/callbacks.py", line 794, in _grad_norm
    param_norms = torch.stack(
RuntimeError: stack expects a non-empty TensorList
```

The function `_grad_norm()` could possibly be updated to return a zero norm if the parameter gradients are empty, but I am unsure if this is meaningful or if it would produce unintended results from the SAM algorithm

``` python
def _grad_norm(self, optimizer: Optimizer) -> torch.Tensor:
    grad_norm_list = [
        (self._norm_weights(p) * p.grad).norm()
        for p in self._get_params(optimizer)
        if isinstance(p.grad, torch.Tensor)
    ]
    if grad_norm_list != []:
        param_norms = torch.stack(grad_norm_list)
    else:
        param_norms = torch.Tensor(0.)
    return param_norms.norm()
```

### Steps to reproduce the problem

Run the example: `./examples/tasks/multitask/three_datasets.py` with this updated trainer:
```python
trainer = pl.Trainer(
    fast_dev_run=100,
    logger=False,
    enable_checkpointing=False,
    callbacks=[cb.GradientCheckCallback(), cb.SAM()],
)
```

### Specifications

Latest and greatest.

**Comment by laserkelvin** [comment]
My gut feeling is that if they were empty to begin with, then it wouldn't affect the end result since they weren't used. I think it's worth a try

**Comment by melo-gonzo** [comment]
I've found the problem and am working on a fix now, will submit a PR soon. It is related to not having a guarantee of each dataset not having the same number of samples, and also not having each dataset present in a batch. Because of this, a loss value for every dataset is not always present, and the gradients of parameters associated with a loss head which was not used are None. It stems from the fact that we iterate over every optimizer regardless of if its associated task head was used or not in the `MultiTaskLitModule`.

**Comment by melo-gonzo** [comment]
Fixed in #192 and #193.
